### PR TITLE
test: settings: functional: fix missing test_main

### LINKED ISSUE #19582: zephyr_library: missing 'kernel-mode' vs 'app-mode' documentation
Not sure where this should be documented, probably as a comment in `cmake/extensions/cmake`?

From: `tests/subsys/settings/functional: make common code a zephyr_library()` #19531, see also references there.
```
zephyr_library_named(  settings_func_test)
add_dependencies(      settings_func_test offsets_h)
```

@mbolivar :

> Why is this `add_dependencies()` required?
> 
> I see this in the top level zephyr/CMakeLists.txt:
> 
> foreach(zephyr_lib ${ZEPHYR_LIBS_PROPERTY})
>   add_dependencies(${zephyr_lib} ${OFFSETS_H_TARGET})
> endforeach()
> So it looks like zephyr_library_named() should be taking care of this:

@SebastianBoe 
> From: "Bøe, Sebastian" <Sebastian.Boe@nordicsemi.no>
> Date: Thursday, 3 October 2019 at 08:12
> To: @marc-hb , @mbolivar , @nvlsianpu 
> Subject: Re: [zephyrproject-rtos/zephyr] tests/subsys/settings/functional: make common code a zephyr_library() (#19531)
> 
> The build system runs in two "modes", or phases, kernel-mode,
> and app-mode. Scripts executed directly or in-directly by boilerplate.cmake
> is in kernel-mode, everything else is in app-mode.
> 
> It's not supported to create zephyr libraries in app-mode as we are doing here,
> but this is not documented anywhere AFAIK so therefore us needing
> to do workarounds like this is a "lack-of-documentation bug" and not a "design bug".
> 
> We use the API in this unsupported way in this case because it happens to do roughly
> (with the exception of needing to manually add_dependencies) what we need to do
> with minimal build script boilerplate.
> 
> The supported way of doing nearly the same thing would be for the app build script to
> add a zephyr_extra_module because west modules are run in kernel-mode and can use
> the zephyr_library API in a supported way. But I decided not to do it in this case because
> creating a zephyr module for this seemed like overkill.

### LINKED ISSUE #42396: ztest: week test_main() is promoted over given testsuite's test_main() if the testsuite uses own library
**Describe the bug**

When I build and run affected tests for, for instance  "system.settings.functional.nvs" for native_posix board, I only obtain this ouput : 
```
*** Booting Zephyr OS build v2.7.99-2256-g8aeed7eaba71  ***
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

It looks like no tests are run.
While debugging this unexpected behaviour, I found that the `test_main` function from `settings_basic_test.c` is not called but it calls the weaked one from `ztest.c` : `void __weak test_main(void)`. That is why nothing is done when I run these tests.
Settings tests seems to be the only ones implementing a zephyr_library which call the ztest framework. It could be the reason why the function `test_main` is not called but I did not dig further.

And indeed the exapmle testsuite uses own library which includes the test_main(); The function is not linked in as the week symbol is found already (it is how the linker works).

This issue is caused by introduction of the week test_main implementation: https://github.com/zephyrproject-rtos/zephyr/pull/37192

**To Reproduce**
`west build -b native_posix tests/subsys/settings/functional/nvs/ -p -t run`

**Expected behavior**
Correctly run all the tests compiled. For instance, it can generate this output : 

```
*** Booting Zephyr OS build v2.7.99-2256-g8aeed7eaba71  ***
Running test suite settings_test_suite
===================================================================
START - test_clear_settings
 PASS - test_clear_settings in 0.0 seconds
===================================================================
START - test_support_rtn
 PASS - test_support_rtn in 0.0 seconds
===================================================================
START - test_register_and_loading
[00:00:00.000,000] <inf> fs_nvs: 4 Sectors of 4096 bytes

[...]

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

**Impact**
While running the tests with twister, it could consider that tests are all passed but nothing is done.

**Environment (please complete the following information):**
- zephyr-rtos 3.0.0-rc1

### LINKED ISSUE #42723: tests: kernel.condvar: child thread is not running
**Describe the bug**
The child thread is not actually running, main thread priority is -1, the two child threads have priorities 0 and 1, respectively. 
```
#define PRIO_WAIT (CONFIG_ZTEST_THREAD_PRIORITY + 2)
#define PRIO_WAKE (CONFIG_ZTEST_THREAD_PRIORITY + 1)
```
```
void test_condvar_wait_forever_wake(void)
{
	woken = 1;
	timeout = K_TICKS_FOREVER;


	k_thread_create(&condvar_tid, stack_1, STACK_SIZE,
			condvar_wait_wake_task, &timeout, NULL, NULL,
			PRIO_WAIT, K_USER | K_INHERIT_PERMS, K_NO_WAIT);

	/* giving time for the condvar_wait_wake_task to execute */
	k_yield();

	k_thread_create(&condvar_wake_tid, condvar_wake_stack, STACK_SIZE,
			condvar_wake_task, &woken, NULL, NULL,
			PRIO_WAKE, K_USER | K_INHERIT_PERMS, K_MSEC(1));

	/* giving time for the condvar_wake_task
	 * and condvar_wait_wake_task to execute
	 */
	k_yield();

	k_thread_abort(&condvar_wake_tid);
	k_thread_abort(&condvar_tid);
}
```
So **k_yield()** does not cause the child thread to execute.

**Environment (please complete the following information):**
 - OS: Windows
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: v2.7.0-ncs1

### PR DESCRIPTION
**Pull request by yperess**
The functional tests for settings subsystem regressed with the
introduction of the weak test_main function. This was caused by
Issue #19582 since the common code was being linked as a library.

Update the logic of the common code to directly add the sources
to the `app` target. This resolved both Issue #19582 and #42396

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/42396
Fixes #19582

Signed-off-by: Yuval Peress <peress@google.com>

### CODE REVIEW

**Comment by nvlsianpu** [comment]
@yperess 
Hi,
I think more expected is to provide a patch which will allow main_test() function in tests-suite's own library. Sorry that I didn't describe that expectation in the issue.

**Comment by yperess** [comment]
> @yperess Hi, I think more expected is to provide a patch which will allow main_test() function in tests-suite's own library. Sorry that I didn't describe that expectation in the issue.

I see what you're saying, but I don't think that's possible. The linking order matters here. One alternative (which I don't mind changing this PR to) is to inverse the relationship. Basically add `settings_basic_test.c` as the source for `app` and have the various tests (`fcb`, `file`, and `nvs`) create the library which is linked into `app`

**Comment by yperess** [comment]
https://stackoverflow.com/questions/51656838/attribute-weak-and-static-libraries/51970423#51970423 actually gives a very good description of what's going on under the hood.

**Comment by yperess** [comment]
Ping

**Review by tejlmand — APPROVED** [review]

**Review by gopiotr — CHANGES_REQUESTED** [review]
Your commit fix only those two tests:
```
tests/subsys/settings/functional/nvs
tests/subsys/settings/functional/fcb
```
During works connected with this PR: https://github.com/zephyrproject-rtos/zephyr/pull/42482 I discovered, that exactly the same problems occur in following tests:
```
tests/subsys/settings/fcb/raw
tests/subsys/settings/littlefs/base64
tests/subsys/settings/littlefs/raw
tests/subsys/settings/nvs/raw
```
Please consider of fixing those bugs in all those problematic tests from `tests/subsys/settings/` directory in this PR

**Comment by yperess** [comment]
This PR now fixes all the tests under `tests/subsys/settings/`

File: tests/subsys/settings/fcb/raw/prj_native_posix.conf [hunk 1]
```diff
@@ -1,5 +1,4 @@
 CONFIG_ZTEST=y
-CONFIG_STDOUT_CONSOLE=y
```
Comment (tejlmand) [inline tests/subsys/settings/fcb/raw/prj_native_posix.conf:2]: Question, did you investigate whether [testcase.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/3631fc4ed87d93cceb4affeb2843ff91897ce430/tests/subsys/settings/fcb/raw/testcase.yaml) should be updated and split in two parts.

One for the native posix platforms, and one for other platforms where you can then add:
```
extra_configs:
- CONFIG_STDOUT_CONSOLE=y
```

I assume `CONFIG_STDOUT_CONSOLE=y` was added originally for a reason.

Comment (yperess) [reply tests/subsys/settings/fcb/raw/prj_native_posix.conf:2]: I did not, seemed unrelated. I went ahead and added that as a separate commit on this PR to clean it up.

**Review by tejlmand — APPROVED** [review]
Looks good, Thanks for the last cleanup commit :+1:

**Comment by yperess** [comment]
> Looks good, Thanks for the last cleanup commit +1

Accidentally left in the `native_posix` platforms in `littlefs/raw`. Fixed it in the latest patch

**Review by gopiotr — APPROVED** [review]

**Review by nvlsianpu — COMMENTED** [review]
Checked that it actually works well.
post merge. +1

**Pull request merged**